File: chronicle-pc/pipeline/chronicle_pipeline/export.py
"""Export Chronicle data (chronosflow format)."""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from pathlib import Path

from .config import ensure_config
from .entries import load_all_entries
from .paths import atomic_write_text, resolve_chronicle_dir

log = logging.getLogger("chronicle.export")


def export_chronosflow(root: Path, out_path: Path) -> Path:
    """
    Chronosflow: a portable JSON bundle of entries + brain summaries.
    """
    cfg = ensure_config(root)
    entries = [e.model_dump(mode="json") for e in load_all_entries(root, fallback_tz=cfg.timezone)]
    brain: dict = {}
    for name in ("graph.json", "tags.json", "prompts.json"):
        p = root / "brain" / name
        if p.is_file():
            brain[name] = json.loads(p.read_text(encoding="utf-8"))

    bundle = {
        "format": "chronosflow",
        "version": 1,
        "exported": datetime.now(timezone.utc).replace(microsecond=0).isoformat(),
        "entries": entries,
        "brain": brain,
    }
    text = json.dumps(bundle, indent=2, ensure_ascii=False) + "\n"
    atomic_write_text(out_path, text)
    return out_path


def run_export(
    chronicle_dir: Path | str | None = None,
    *,
    format: str = "chronosflow",
    path: str | Path | None = None,
) -> dict:
    root = resolve_chronicle_dir(chronicle_dir)
    fmt = (format or "chronosflow").lower()
    if fmt != "chronosflow":
        raise ValueError(f"unsupported export format: {format} (supported: chronosflow)")

    out = Path(path) if path else root / f"chronicle-export-{datetime.now().strftime('%Y%m%d')}.chronosflow.json"
    out = out.expanduser().resolve()
    export_chronosflow(root, out)
    log.info("Exported chronosflow → %s", out)
    return {"format": fmt, "path": str(out)}

File: chronicle-pc/server/src/export.rs
//! chronosflow export bundle (export.py).

use std::path::Path;

use serde_json::{json, Value};

use crate::entries as store;
use crate::errors::ChronicleError;
use crate::paths::{atomic_write_json, resolve_chronicle_dir};

pub const SUPPORTED_FORMATS: [&str; 1] = ["chronosflow"];

pub fn run_export(dir: Option<&Path>, format: Option<&str>, out_path: Option<&Path>) -> Result<Value, ChronicleError> {
    let fmt = format.unwrap_or("chronosflow").to_lowercase();
    if !SUPPORTED_FORMATS.contains(&fmt.as_str()) {
        return Err(ChronicleError::msg(format!(
            "unsupported export format: {fmt} (supported: chronosflow)"
        )));
    }
    let root = resolve_chronicle_dir(dir)?;
    let entries: Vec<Value> = store::load_all_entries(&root)?
        .iter()
        .map(|e| e.to_api_value())
        .collect();
    let mut brain = serde_json::Map::new();
    for name in ["graph.json", "tags.json", "prompts.json"] {
        let p = root.join("brain").join(name);
        if let Ok(v) = crate::paths::read_json(&p) {
            brain.insert(name.to_string(), v);
        }
    }
    let payload = json!({
        "format": "chronosflow",
        "version": 1,
        "exported": crate::timeutil::now_iso(),
        "entries": entries,
        "brain": Value::Object(brain),
    });
    let target = out_path.map(Path::to_path_buf).unwrap_or_else(|| {
        root.join(format!(
            "chronicle-export-{}.chronosflow.json",
            chrono::Local::now().format("%Y%m%d")
        ))
    });
    atomic_write_json(&target, &payload)?;
    Ok(json!({ "format": fmt, "path": target.to_string_lossy() }))
}
